go/pumiceDB/client/PumiceDBClient.go

```diff
@@ -93,132 +107,86 @@ func getPmdbReq(reqArgs *PmdbReqArgs) (unsafe.Pointer, int64) {
+}
 
-	enc := gob.NewEncoder(&rBytes)
-	err = enc.Encode(reqArgs.ReqED)
-	if err != nil {
-		return nil, err
+// Encode the user's portion of the request
+func (r *PmdbReqArgs) encodeRequestPayload() error {
+	// If 'request' is present then the contents are already here
+	if r.request != nil { // Already encoded
+		return nil
 	}
 
-	reqArgs.ReqByteArr = rBytes.Bytes()
-
-	//Convert to unsafe pointer (void * for C function)
-	eData, reqLen := getPmdbReq(reqArgs)
+	if r.IRequest == nil {
+		return fmt.Errorf("IRequest is not present")
+	}
 
-	//Typecast the encoded key to char*
-	ekey := (*C.char)(eData)
-	getResC := (C.int)(reqArgs.GetResponse)
+	var b bytes.Buffer
 
-	//Perform the write
-	return obj.writeKV(reqArgs.Rncui, ekey, reqLen, getResC,
-		reqArgs.ReplySize)
-}
+	err := gob.NewEncoder(&b).Encode(r.IRequest)
+	if err == nil {
+		r.request = b.Bytes()
+	}
 
-//WriteEncoded
-/*
-WriteEncoded allows client to pass the encoded KV struct for writing
-*/
-func (obj *PmdbClientObj) WriteEncoded(reqArgs *PmdbReqArgs) (unsafe.Pointer,
-	error) {
-	//Convert it to unsafe pointer (void * for C function)
-	eData := unsafe.Pointer(&reqArgs.ReqByteArr[0])
-	reqLen := int64(len(reqArgs.ReqByteArr))
-	eReq := (*C.char)(eData)
-	getResC := (C.int)(reqArgs.GetResponse)
-	return obj.writeKV(reqArgs.Rncui, eReq, reqLen,
-		getResC,
-		reqArgs.ReplySize)
+	return err
 }
 
-func (obj *PmdbClientObj) WriteEncodedAndGetResponse(reqArgs *PmdbReqArgs) error {
-	var replySize int64
-	var wr_err error
-	var replyB unsafe.Pointer
-
-	//Convert to unsafe pointer (void * for C function)
-	eData := unsafe.Pointer(&reqArgs.ReqByteArr[0])
-	reqLen := int64(len(reqArgs.ReqByteArr))
-	eReq := (*C.char)(eData)
-	getResC := (C.int)(reqArgs.GetResponse)
-
-	replyB, wr_err = obj.writeKV(reqArgs.Rncui, eReq,
-		reqLen, getResC, &replySize)
-	if wr_err != nil {
-		return wr_err
+// Encode the user's data along w/ the pmdb request
+func (r *PmdbReqArgs) encodeRequest() (unsafe.Pointer, int64, error) {
+	if r.request == nil {
+		return nil, 0, fmt.Errorf("r.request is nil")
 	}
 
-	if replyB != nil {
-		bytes_data := C.GoBytes(unsafe.Pointer(replyB), C.int(replySize))
-		buffer := bytes.NewBuffer(bytes_data)
-		*reqArgs.Response = buffer.Bytes()
+	pmdbReq := &PumiceDBCommon.PumiceRequest {
+		ReqType: PumiceDBCommon.APP_REQ,
+		ReqPayload: r.request,
 	}
-	// Free the buffer allocated by the C library
-	C.free(replyB)
-	return nil
-}
 
-//Read the value of key on the client
-func (obj *PmdbClientObj) Read(reqArgs *PmdbReqArgs) error {
+	var len int64
+	ptr, err := PumiceDBCommon.Encode(pmdbReq, &len)
 
-	var replySize int64
-	var rd_err error
-	var replyB unsafe.Pointer
-	var requestBytes bytes.Buffer
-	var err error
+	return ptr, len, err
+}
 
-	enc := gob.NewEncoder(&requestBytes)
-	err = enc.Encode(reqArgs.ReqED)
-	if err != nil {
-		return err
-	}
+func (r *PmdbReqArgs) getObjectID(o *C.pmdb_obj_id_t)  error {
+	rncuiStrC := GoToCString(r.Rncui)
+	defer FreeCMem(rncuiStrC)
```

Comment:
These helper methods, `FreeCMem()`,  `GoToCString()`, etc should probably be removed.  They can be directly replaced by the intrinsic CGO methods which they wrap.   If there's some reason for keeping them, then the methods in this file should be made private OR moved to `PumiceDBCommon` (though that seems like overkill).  For me calling `C.free(unsafe.Pointer(cstring))` is preferred.